### ISSUE TITLE
HTML bundler plugin fails to resolve source files

### ISSUE THREAD

**Issue by Mrtenz** [MetaMask/metamask-extension#27831]
Sometimes when running Webpack, the HTML bundler plugin fails to resolve certain source files, like `scripts/load/background.ts`, and `scripts/load/ui.ts`, even though these files exist. This seems to be due to some kind of caching issue, as it can be resolved when clearing `node_modules/.cache`, and only happens when the cache is already there.

I don't have a consistent way to reproduce it, but it sometimes happens after running `yarn webpack -w --type flask`, killing the process, and running the same command again.

```
🦊 Running development build…
ERROR in [initial]
 HTML Bundler Plugin  Failed to execute the function.
Source file: './app/background.html'
./node_modules/html-bundler-webpack-plugin/src/Plugin/Messages/Exception.js:110
  throw new PluginException(message);
  ^

PluginException: 
 HTML Bundler Plugin  Can't resolve scripts/load/background.ts in the file background.html

    at resolveException (./node_modules/html-bundler-webpack-plugin/src/Plugin/Messages/Exception.js:110:9)
    at Resolver.require (./node_modules/html-bundler-webpack-plugin/src/Plugin/Resolver.js:262:7)
    at ./app/background.html:1:132
    at Script.runInContext (node:vm:148:12)
    at VMScript.exec (./node_modules/html-bundler-webpack-plugin/src/Common/VMScript.js:39:27)
    at HtmlBundlerPlugin.renderModule (./node_modules/html-bundler-webpack-plugin/src/Plugin/AssetCompiler.js:1314:27)
    at HtmlBundlerPlugin.renderManifest (./node_modules/html-bundler-webpack-plugin/src/Plugin/AssetCompiler.js:889:26)
    at Hook.eval [as call] (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at Compilation.getRenderManifest (./node_modules/webpack/lib/Compilation.js:4689:36)
    at ./node_modules/webpack/lib/Compilation.js:4709:22

ERROR in [initial]
 HTML Bundler Plugin  Failed to execute the function.
Source file: './app/home.html'
./node_modules/html-bundler-webpack-plugin/src/Plugin/Messages/Exception.js:110
  throw new PluginException(message);
  ^

PluginException: 
 HTML Bundler Plugin  Can't resolve scripts/load/ui.ts in the file home.html

    at resolveException (./node_modules/html-bundler-webpack-plugin/src/Plugin/Messages/Exception.js:110:9)
    at Resolver.require (./node_modules/html-bundler-webpack-plugin/src/Plugin/Resolver.js:262:7)
    at ./app/home.html:1:896
    at Script.runInContext (node:vm:148:12)
    at VMScript.exec (./node_modules/html-bundler-webpack-plugin/src/Common/VMScript.js:39:27)
    at HtmlBundlerPlugin.renderModule (./node_modules/html-bundler-webpack-plugin/src/Plugin/AssetCompiler.js:1314:27)
    at HtmlBundlerPlugin.renderManifest (./node_modules/html-bundler-webpack-plugin/src/Plugin/AssetCompiler.js:889:26)
    at Hook.eval [as call] (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at Compilation.getRenderManifest (./node_modules/webpack/lib/Compilation.js:4689:36)
    at ./node_modules/webpack/lib/Compilation.js:4709:22

ERROR in [initial]
 HTML Bundler Plugin  Failed to execute the function.
Source file: './app/notification.html'
./node_modules/html-bundler-webpack-plugin/src/Plugin/Messages/Exception.js:110
  throw new PluginException(message);
  ^

PluginException: 
 HTML Bundler Plugin  Can't resolve scripts/load/ui.ts in the file notification.html

    at resolveException (./node_modules/html-bundler-webpack-plugin/src/Plugin/Messages/Exception.js:110:9)
    at Resolver.require (./node_modules/html-bundler-webpack-plugin/src/Plugin/Resolver.js:262:7)
    at ./app/notification.html:1:1383
    at Script.runInContext (node:vm:148:12)
    at VMScript.exec (./node_modules/html-bundler-webpack-plugin/src/Common/VMScript.js:39:27)
    at HtmlBundlerPlugin.renderModule (./node_modules/html-bundler-webpack-plugin/src/Plugin/AssetCompiler.js:1314:27)
    at HtmlBundlerPlugin.renderManifest (./node_modules/html-bundler-webpack-plugin/src/Plugin/AssetCompiler.js:889:26)
    at Hook.eval [as call] (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at Compilation.getRenderManifest (./node_modules/webpack/lib/Compilation.js:4689:36)
    at ./node_modules/webpack/lib/Compilation.js:4709:22

ERROR in [initial]
 HTML Bundler Plugin  Failed to execute the function.
Source file: './app/popup.html'
./node_modules/html-bundler-webpack-plugin/src/Plugin/Messages/Exception.js:110
  throw new PluginException(message);
  ^

PluginException: 
 HTML Bundler Plugin  Can't resolve scripts/load/ui.ts in the file popup.html

    at resolveException (./node_modules/html-bundler-webpack-plugin/src/Plugin/Messages/Exception.js:110:9)
    at Resolver.require (./node_modules/html-bundler-webpack-plugin/src/Plugin/Resolver.js:262:7)
    at ./app/popup.html:1:960
    at Script.runInContext (node:vm:148:12)
    at VMScript.exec (./node_modules/html-bundler-webpack-plugin/src/Common/VMScript.js:39:27)
    at HtmlBundlerPlugin.renderModule (./node_modules/html-bundler-webpack-plugin/src/Plugin/AssetCompiler.js:1314:27)
    at HtmlBundlerPlugin.renderManifest (./node_modules/html-bundler-webpack-plugin/src/Plugin/AssetCompiler.js:889:26)
    at Hook.eval [as call] (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at Compilation.getRenderManifest (./node_modules/webpack/lib/Compilation.js:4689:36)
    at ./node_modules/webpack/lib/Compilation.js:4709:22

ERROR in [initial]
 HTML Bundler Plugin  Failed to execute the function.
Source file: './app/trezor-usb-permissions.html'
./node_modules/html-bundler-webpack-plugin/src/Plugin/Messages/Exception.js:110
  throw new PluginException(message);
  ^

PluginException: 
 HTML Bundler Plugin  Can't resolve vendor/trezor/usb-permissions.js in the file trezor-usb-permissions.html

    at resolveException (./node_modules/html-bundler-webpack-plugin/src/Plugin/Messages/Exception.js:110:9)
    at Resolver.require (./node_modules/html-bundler-webpack-plugin/src/Plugin/Resolver.js:262:7)
    at ./app/trezor-usb-permissions.html:1:1211
    at Script.runInContext (node:vm:148:12)
    at VMScript.exec (./node_modules/html-bundler-webpack-plugin/src/Common/VMScript.js:39:27)
    at HtmlBundlerPlugin.renderModule (./node_modules/html-bundler-webpack-plugin/src/Plugin/AssetCompiler.js:1314:27)
    at HtmlBundlerPlugin.renderManifest (./node_modules/html-bundler-webpack-plugin/src/Plugin/AssetCompiler.js:889:26)
    at Hook.eval [as call] (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at Compilation.getRenderManifest (./node_modules/webpack/lib/Compilation.js:4689:36)
    at ./node_modules/webpack/lib/Compilation.js:4709:22

🦊 MetaMask – development compiled with 5 errors
🦊 Watching for changes…
```

**Comment by webdiscus** [comment]
@Mrtenz 

No wonder, after aborting (ctrl+c) the long build process, the cache is corrupted and may contains invalid values.

I have tested with complete build process and it's works fine with caching.

## Solution

After manually aborting the build process, the cache must be manually deleted.

## Recommend 

You can use the `cacheDirectory` option to define the `.cache` dir outer the `node_modules/`:

```js
cache: {
    type: 'filesystem',
    cacheDirectory: path.join(__dirname, '.cache'),
    ...
},
```

**Comment by Mrtenz** [comment]
> No wonder, after aborting (ctrl+c) the long build process, the cache is corrupted and may contains invalid values.

Note that I'm running in watch mode. There is no way to exit the process gracefully, it seems. I wouldn't expect the cache to become invalid in this case.

**Comment by webdiscus** [comment]
> > No wonder, after aborting (ctrl+c) the long build process, the cache is corrupted and may contains invalid values.
> 
> Note that I'm running in watch mode. There is no way to exit the process gracefully, it seems. I wouldn't expect the cache to become invalid in this case.

Ah, yes, you are right. I was confused because I use `webpack watch` instead of `webpack -w`.
I confirm, after stopping the watch/serve mode occurs an error (I have another error message but...) and works only after clearing cache.

I will research this case to find & fix the problem.

**Comment by webdiscus** [comment]
@Mrtenz 

the issue is fixed in the `v4.1.2`.
Can you please check it.

**Comment by webdiscus** [comment]
@davidmurdoch @Mrtenz 

can you please confirm whether the `v4.1.2` fixes the issue?
